### PR TITLE
storage daemon: fix crash on volume swap

### PR DESCRIPTION
**Pull request by arogge**
Fix a problem introduced in Bareos 22 with PR #1271 where the SD would crash on a volume swap.

This PR also introduces a testing backend to make it easier to test autochangers. However, a change in the Device interface is required for this (so we shouldn't backport that).

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)

##### General
- [x] PR name is meaningful
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
- [x] Check backport line
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Separate commit for CHANGELOG.md ("update CHANGELOG.md"). The PR number is correct.

##### Source code quality

- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR
- [x] `bareos-check-sources --since-merge` does not report any problems

##### Tests

- [x] Decision taken that a test is required (if not, then remove this paragraph)
- [x] The choice of the type of test (unit test or systemtest) is reasonable
- [x] Testname matches exactly what is being tested
- [x] On a fail, output of the test leads quickly to the origin of the fault

### CODE REVIEW

File: core/src/stored/vol_mgr.cc [hunk 1]
```diff
@@ -480,17 +480,20 @@ VolumeReservationItem* reserve_volume(DeviceControlRecord* dcr,
...
-        dcr->dev->swap_dev = vol->dev;        /* remember to get this vol */
-        dcr->dev->SetLoad();                  /* then reload on our drive */
-        vol->dev->vol = NULL; /* remove volume from other drive */
-        vol->dev = dcr->dev;  /* point the Volume at our drive */
-        dcr->dev->vol = vol;  /* point our drive at the Volume */
+        {
+          auto dev = dcr->dev;
+          dcr->SetDev(vol->dev);                /* temp point to other dev */
+          slot = GetAutochangerLoadedSlot(dcr); /* get slot on other drive */
+          dcr->SetDev(dev);                     /* restore dev */
```
Comment (pstorz) [inline core/src/stored/vol_mgr.cc:487]:
```suggestion
          auto original_dcr_dev = dcr->dev;
          dcr->SetDev(vol->dev);                /* temp point to other dev */
          slot = GetAutochangerLoadedSlot(dcr); /* get slot on other drive */
          dcr->SetDev(original_dcr_dev);                     /* restore dev */
```

**Review by pstorz — CHANGES_REQUESTED** [review]
Good work! Impressive, how simple now it is to add a new backend device.

File: core/src/stored/backends/CMakeLists.txt [hunk 2]
```diff
@@ -67,3 +67,14 @@ if(HAVE_DARWIN_OS)
   target_link_libraries(bareossd-fifo bareos bareossd)
   target_link_libraries(bareossd-gentape bareos bareossd)
 endif()
+
+if(HAVE_DYNAMIC_SD_BACKENDS)
+  add_library(bareossd-autochanger_test MODULE)
+  set_property(
+    TARGET bareossd-autochanger_test PROPERTY POSITION_INDEPENDENT_CODE ON
```
[the remark this answers is not in the document]
Comment (arogge) [reply core/src/stored/backends/CMakeLists.txt:74]: I guess we can remove it in this case.
Basically, as we have it enabled in other parts of the code and it is the default for SHARED and MODULE targets, we only have to set it on some special occassions.

If you build without dynamically loadable backends, these will be linked statically into the libbareossd shared library. Therefore the backends needs to have PIC enabled.

Maybe we could globabally enable `CMAKE_POSITION_INDEPENDENT_CODE` in `core/CMakeLists.txt` so it is the default.

File: core/src/stored/vol_mgr.cc (hunk 1 above)
Comment (arogge) [reply core/src/stored/vol_mgr.cc:487]: Good idea. I'll rename it.

**Review by pstorz — APPROVED** [review]
Good work!

**Pull request merged**